packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -68,330 +475,366 @@ class FocusNode extends ChangeNotifier {
+    final Set<FocusNode> oldFocusPath = oldPrimaryFocus?.ancestors?.toSet() ?? <FocusNode>{};
+    child._parent?._removeChild(child);
+    _children.add(child);
+    child._parent = this;
+    child._updateManager(_manager);
+    if (oldPrimaryFocus != null) {
+      final Set<FocusNode> newFocusPath = _manager?._currentFocus?.ancestors?.toSet() ?? <FocusNode>{};
+      // Nodes that will no longer be focused need to be marked dirty.
+      for (FocusNode node in oldFocusPath.difference(newFocusPath)) {
+        node._markAsDirty();
+      }
+      // If the node used to have focus, make sure it keeps it's old primary
+      // focus when it moves.
+      oldPrimaryFocus.requestFocus();
+    }
+  }
+
+  /// Called by the _host_ [StatefulWidget] to attach a [FocusNode] to the focus tree.
   ///
-  /// This method is safe to call regardless of whether this node has ever
-  /// requested focus.
-  void unfocus() {
-    _parent?._resignFocus(this);
-    assert(_parent == null);
-    assert(_manager == null);
+  /// In order to attach a [FocusNode] to the focus tree, call [attach] when the
+  /// node is ready to be added to the focus tree, typically from the
+  /// [StatefulWidget]'s [State.initState] method.
+  ///
+  /// If the focus node in the host widget is swapped out, the new node will
+  /// need to be attached. [FocusAttachment.detach] should be called on the old
+  /// node, and then [attach] called on the new node. This typically happens in
+  /// the [State.didUpdateWidget] method.
+  @mustCallSuper
+  FocusAttachment attach(BuildContext context, {FocusOnKeyCallback onKey}) {
+    _context = context;
+    _onKey = onKey;
+    _attachment = FocusAttachment._(this);
+    return _attachment;
   }
 
   @override
   void dispose() {
     _manager?._willDisposeFocusNode(this);
-    _parent?._resignFocus(this);
-    assert(_parent == null);
-    assert(_manager == null);
+    _attachment?.detach();
     super.dispose();
   }
 
+  @mustCallSuper
   void _notify() {
+    if (_parent == null) {
+      // no longer part of the tree, so don't notify.
+      return;
+    }
+    if (hasPrimaryFocus) {
+      _setAsFocusedChild();
+    }
     notifyListeners();
   }
 
+  /// Requests the primary focus for a [node], which will also give
+  /// focus to its [ancestors].
+  ///
+  /// If called without a node, request focus for this node.
+  ///
+  /// If the given [node] is a [FocusScopeNode] and that focus scope node has a
+  /// non-null [focusedChild], then request the focus for the focused child.
+  /// This process is recursive and continues until it encounters either a focus
+  /// scope node with a null focused child or an ordinary (non-scope)
+  /// [FocusNode] is found.
+  ///
+  /// The node is notified that it has received the primary focus in a
+  /// microtask, so notification may lag the request by up to one frame.
+  void requestFocus([FocusNode node]) {
+    if (node != null) {
+      if (node._parent == null) {
```

Comment:
Why do we reparent the node to ourselves if it is unowned?

Reply:
Judging from other similar methods in this PR, this fact probably just needs to be documented in the doc comment?